### ISSUE TITLE
Libtorch C++ multiple GPU performance slower than single GPU

### ISSUE THREAD

**Issue by bootpin** [pytorch/pytorch#40581]
## 🐛 Bug

Using the data_parallel C++ interface results in code that is much slower on multiple GPUs than on a single GPU.  In addition, the GPU utilization is less than 10% with muliple GPUs compared to over 96% with a single GPU.

## To Reproduce

 Implement a non-trivial model (e.g. ResNet50 or SlowFast) using the libtorch c++ interface (I used libtorch-win-shared-with-deps-1.5.0.zip + cuda 10.1)

Add mulitiple GPU support via torch::nn::parallel::data_parallel

Time training runs with a single GPU and with multiple GPUs 

## Expected behavior

Runs with multiple GPUs should be faster than runs on a single GPU.  Definitely not *much* slower.

## Environment

 - PyTorch Version (e.g., 1.0): LibTorch 1.5.0 pre-built library
 - OS (e.g., Linux): Windows 7
 - How you installed PyTorch (`conda`, `pip`, source): N/A
 - Build command you used (if compiling from source): 
 - Python version: N/A
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: 2x GTX 1080
 - Any other relevant information:

## Additional context

This is in line with what @dmagee reported in #18837. Looking though the code, it appears as if replicas of the modules are cloned and deleted on every iteration of training. Is there a way to use data_parallel and avoid this overhead?

**Comment by mruberry** [comment]
@ngimel

**Comment by ngimel** [comment]
The recommended way to to use DistributedDataParallel instead of DataParallel https://pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel. I don't know the status of DistributedDataParallel support with c++ apis though. 
That said, what you are observing is strange, and just replication overhead should not be enough to explain the slowdown that you are seeing. In python the models are replicated too, and yet typically there is a speed-up with DataParallel.

**Comment by bootpin** [comment]
@ngimel, thanks for the update.  
The model I am trying to train has about 250 modules (Conv3d, BatchNorm3d etc.). These get registered (via register_module) whenever **reset()** is called, and **reset()** gets called every training iteration.  I am trying to see how I can create a copy of my model on each GPU only once and perform weight updates correctly.  Running the parallel forward passes is no problem but I don't yet know the libtorch code well enough to merge gradients from multiple GPUs on the backward passes.  I'm taking a look at the **replicate_grad_edges** code to see if I can get any insights.

**Comment by glaringlee** [comment]
cc @mrshenli as well

**Comment by bootpin** [comment]
I modified my model to create replicas only on the first call to _**clone**_ for each GPU.  Subsequent calls to _**clone**_ simply return cached model pointers after updating parameters.  The results in terms of performance are night and day.  I now get about 1.2x speed up with two GPUs.  Not great but way better than the roughly 0.3x I was seeing previously.

I am updating the parameters by calling **_weight.copy__** and **_bias.copy__** but I'm not sure if this is all that is required.  Training loss decreases as expected and so multi GPU training appears to be happening.

**Comment by VitalyFedyunin** [comment]
@bootpin do you feel we can close this issue?

**Comment by linjiapro** [comment]
I am seeing exact same issue: training on 2 GPUS is slower than 1 GPU. The issue might be the clone class's reset() function gets called for every batch. I don't think we should hack in a way that reset function reset to a cached version of parameters, though this will make training quicker, the reason is that if we do this, we change the behavior of the clone; this will create bug when clone is called elsewhere.

A proper fix might be that for data_parallel, instead of creating deep clone every time, data_parallel needs to do caching on GPU, instead of invoking clone API. However, as @ngimel mentioned, DDP is the future. So a better way is just to discard data_parallel, and implement DDP on C++?

**Comment by bootpin** [comment]
Yes we should move to DistributedDataParallel. My hack allowed me to scale training to multiple GPUs but would not be an acceptable addition to the code base.  I however cannot find a C++ implementation of DistributedDataParallel in the 1.6 release. Is this in the works?

**Comment by ngimel** [comment]
cc @pritamdamania (for ddp) and @glaringlee (for c++ apis). Requests to have ddp in c++ come up regularly, can you guys comment on the plans?

**Comment by pritamdamania87** [comment]
I'm not aware of any immediate plans to support a C++ API for DDP. @ngimel If you have references to additional gh issues/forum posts requesting this feature, it would help us to prioritize a C++ API accordingly.

**Comment by beniz** [comment]
Hi @pritamdamania87 having multi-GPU in C++ on par with its Python counterpart would be a plus for us. I believe the building of full C++ stacks for deep learning is not niche anymore.

**Comment by linjiapro** [comment]
Training using C++ stack will gain more and more traction. C++ offers 2 advantages compared to other languages: 1) its performance is premium, and 2) it is a strong typed, and compiler-based language. IMHO, the second advantage is even more important. As the complexity of software goes up, and the need for performance goes up, C++ shines. Some most sophisticated software in the industry is written in C++ or a similar strong-typed and compiler-based language. For example, EDA design software are mostly written in C++, and the early IOS operation systems is based on C/C++ and Objective C. Deep learning software system will grow more sophisticated and will need more performance. There is no reason that C++ will not become mainstream.

**Comment by bhendrickson** [comment]
One can train on multiple GPUs without using the data_parallel abstraction and the overhead in `.clone()` by making a model for each GPU, using `torch::cuda::nccl::broadcast` to spread weights from the first to the rest, using multiple threads to do forward & backward on each, and using `torch::cuda::nccl::reduce` to combine all gradients back to the first. Besides being lower overhead, this neither requires a model to be cloneable nor have a particular interface for it's 'forward' method.

At least some of the motivation for DistributedDataParallel using separate processes seems to be to get around python's global interpreter lock, which doesn't exist in C++. Although it is nice for the C++ frontend to mirror the Python one, for single computer multi-gpu training, perhaps using separate processes is an unnecessary hassle?

**Comment by bhendrickson** [comment]
At least some of the motivation for DistributedDataParallel using separate processes seems to be to get around python's global interpreter lock, which doesn't exist in C++. Although it is nice for the C++ frontend to mirror the Python one, for single computer multi-gpu training, perhaps using separate processes is an unnecessary hassle?

As @bootpin points out, one can train on multiple GPUs without using the data_parallel abstraction nor the overhead in `.clone()` by making a model for each GPU, and copying weights/gradients around.  Perhaps all that is needed is an example showing how to use this pattern well? Besides being lower overhead, this neither requires a model to be cloneable nor have a particular interface for it's 'forward' method. (I get about 1.7x increase in examples per second doing this on my 2 GPU box training a 183 million weight "bert-like" transformer, using mixed precision so transfering weights/grads as float16. Splitting up responsibility for optimizing the weights across the GPUs improved this very little, but perhaps with more GPUs that would be better.) 

Like @bootpin, I am curious about best practices for broadcasting the weights and reducing the gradients.  Surprisingly, I found `torch::cuda::nccl::broadcast` and `torch::cuda::nccl::reduce` to be slower than stringing together calls of `.to(device)` with `+=` or `=`.

**Comment by xanderdunn** [comment]
@bhendrickson Do you have a pseudo-code example of this? Do you combine your gradients by averaging them, backdrop on one of your models, and then copy those updated weights from that model to the other models on the other GPUs?

**Comment by bhendrickson** [comment]
@xanderdunn Sorta. I actually have two models on the first device, the main model that I broadcast weights from, accumulate gradients to, and run the optimizer on -- and then also the first replica model is on there. The replicas are where I do do the forward/backward passes to make gradients in parallel on all devices. I think having a single model on the first device would be essentially the same and save a bit of memory there. Here is some pseudocode I just made for you, simplified down from my actual code, but no promises it is completely right:

```
    master_parameters = model->parameters();
    for (auto device : devices) {
      replicas.emplace_back(std::dynamic_pointer_cast<typename model_type::Impl>(model->clone()));
      replicas.back()->to(device);
      replica_parameters.push_back(replicas.back()->parameters());
    } 
    AdamOptimizer optimizer(master_parameters)
    for (;;) {
       broadcast_tensors(master_parameters, replica_parameters);
       for (auto& r : replicas) thread_pool.post([&]{compute_loss(r, next_training_batch()).backward()});
       thread_pool.wait();
       reduce_gradients(replica_parameters, master_parameters);
       clear_gradients(replica_parameters);
       optimizer.step();
    }
```

I've actually found `broadcast_tensors` is faster still when I combine all the weights of the master into a single "mega" tensor, use `.to(device)` to move that to each device, and then use something like `for (...) {replica_weight[i][j].variable_data() = copied_mega.narrow(...).view(...)` to assign the specific weights from that "mega" tensor with all the different weights in it. Simularily, I've found it fastest in `reduce_gradients` to consolidate the gradients for each replica into a mega tensor for each replica before moving it all to the same device, adding them together, and then assign to the gradients for each weight in the master model from the correct part of the consolidated gradients tensor using more `narrow(...)` and `view(..)` calls. But also works to what i was doing before, which is loop over the weights or gradients and copy (or copy and accumulate) them separately.

**Comment by beniz** [comment]
FTR, DeepDetect now has a C++ libtorch-based multi-GPU implementation by @Bycob that can serve as a basis for others, see here: https://github.com/jolibrain/deepdetect/pull/1285/files#diff-09269caf32658c7771c16ee95bf497118cf33ea466ddd2539bc3d87478f9d306R811
This is for single machine/multi-GPU, and in our tests it behaves as expected, i.e. ~x4 on 4-GPUs

**Comment by mp3guy** [comment]
> FTR, DeepDetect now has a C++ libtorch-based multi-GPU implementation by @Bycob that can serve as a basis for others, see here: https://github.com/jolibrain/deepdetect/pull/1285/files#diff-09269caf32658c7771c16ee95bf497118cf33ea466ddd2539bc3d87478f9d306R811
> This is for single machine/multi-GPU, and in our tests it behaves as expected, i.e. ~x4 on 4-GPUs

Is this diff the only changes required?

**Comment by Bycob** [comment]
This diff includes more than just multi GPU, as it covers models that are already integrated in DeepDetect.
The multi GPU loop happens in torchlib.cc
- https://github.com/jolibrain/deepdetect/blob/724fbe330607b0d286949b49f5dee779d4ca259d/src/backends/torch/torchlib.cc#L788
Cloning the models, alternatively you can initialize multiple models normally and then sync their weights
- https://github.com/jolibrain/deepdetect/blob/724fbe330607b0d286949b49f5dee779d4ca259d/src/backends/torch/torchlib.cc#L855
Parallel execution on one thread per device
- https://github.com/jolibrain/deepdetect/blob/724fbe330607b0d286949b49f5dee779d4ca259d/src/backends/torch/torchlib.cc#L922
Synchronization of gradients. Backward pass done on master GPU only
- https://github.com/jolibrain/deepdetect/blob/724fbe330607b0d286949b49f5dee779d4ca259d/src/backends/torch/torchlib.cc#L967
Broadcast weights to all instances

This implementation only needs openmp, weights are copied using basic torch API, could probably be improved using NCCL. And that's it.

**Comment by mp3guy** [comment]
Hmm seems pretty straight forward. Sounds like anyone could just retrofit this onto their on training loop C++ frontend code without having to hack torch internals?

**Comment by anqin** [comment]
> @xanderdunn Sorta. I actually have two models on the first device, the main model that I broadcast weights from, accumulate gradients to, and run the optimizer on -- and then also the first replica model is on there. The replicas are where I do do the forward/backward passes to make gradients in parallel on all devices. I think having a single model on the first device would be essentially the same and save a bit of memory there. Here is some pseudocode I just made for you, simplified down from my actual code, but no promises it is completely right:
> 
> ```
>     master_parameters = model->parameters();
>     for (auto device : devices) {
>       replicas.emplace_back(std::dynamic_pointer_cast<typename model_type::Impl>(model->clone()));
>       replicas.back()->to(device);
>       replica_parameters.push_back(replicas.back()->parameters());
>     } 
>     AdamOptimizer optimizer(master_parameters)
>     for (;;) {
>        broadcast_tensors(master_parameters, replica_parameters);
>        for (auto& r : replicas) thread_pool.post([&]{compute_loss(r, next_training_batch()).backward()});
>        thread_pool.wait();
>        reduce_gradients(replica_parameters, master_parameters);
>        clear_gradients(replica_parameters);
>        optimizer.step();
>     }
> ```
> 
> I've actually found `broadcast_tensors` is faster still when I combine all the weights of the master into a single "mega" tensor, use `.to(device)` to move that to each device, and then use something like `for (...) {replica_weight[i][j].variable_data() = copied_mega.narrow(...).view(...)` to assign the specific weights from that "mega" tensor with all the different weights in it. Simularily, I've found it fastest in `reduce_gradients` to consolidate the gradients for each replica into a mega tensor for each replica before moving it all to the same device, adding them together, and then assign to the gradients for each weight in the master model from the correct part of the consolidated gradients tensor using more `narrow(...)` and `view(..)` calls. But also works to what i was doing before, which is loop over the weights or gradients and copy (or copy and accumulate) them separately.

the  pseudocode is pretty good. where can find the implementation codes (or example codes)  about the above  pseudocode.
